packages/common/http/src/context.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+/**
+ * A token used to manipulate and access values stored in HttpContext
+ *
+ * @publicApi
+ */
+export class HttpContextToken<T> {
+  constructor(public readonly defaultValue: T) {}
+}
+
+/**
+ * Http context stores arbitrary user defined values and ensures type safety without
+ * actually knowing the types.
+ *
+ * This context is mutable and is shared between cloned requests.
+ *
+ * @usageNotes
+ *
+ * ### Usage Example
+ *
+ * ```typescript
+ * // inside cache.interceptors.ts
+ * export const IS_CACHE_ENABLED = new HttpContextToken<boolean>(false);
+ *
+ * export class CacheInterceptor implements HttpInterceptor {
+ *
+ *   intercept(req: HttpRequest<any>, delegate: HttpHandler): Observable<HttpEvent<any>> {
+ *     if (req.context.get(IS_CACHE_ENABLED) === true) {
+ *       return ...;
+ *     }
+ *     return delegate.handle(req);
+ *   }
+ * }
+ *
+ * // inside a service
+ *
+ * this.httpClient.get('/api/weather', {
+ *   context: new HttpContext().set(IS_CACHE_ENABLED, true)
+ * }).subscribe(...);
+ * ```
+ *
+ * @publicApi
+ */
+export class HttpContext {
+  private map = new Map<HttpContextToken<unknown>, unknown>();
+
+  /**
+   * Store a value in the context. If a value is already present it gets overwritten.
+   *
+   * @param token The reference to an instance of HttpContextToken
+   * @param value The value
+   *
+   * @returns A reference to itself for easy chaining
+   */
+  set<T>(token: HttpContextToken<T>, value: T): HttpContext {
+    this.map.set(token, value);
+    return this;
+  }
+
+  /**
+   * Retrieve the value associated with the given token
+   *
+   * @param token The reference to an instance of HttpContextToken
+   *
+   * @returns The stored value or default if one is defined
+   */
+  get<T>(token: HttpContextToken<T>): T {
+    return this.map.has(token) ? this.map.get(token) as T : token.defaultValue;
+  }
+
+  /**
+   * Delete the value associated with given token
+   *
+   * @param token The reference to an instance of HttpContextToken
+   * @param value The value
+   *
+   * @returns A reference to itself for easy chaining
+   */
+  delete<T extends HttpContextToken<any>>(token: T): HttpContext {
```

Comment:
There is no need for the generic type parameter here

```suggestion
  delete(token: HttpContextToken<any>): HttpContext {
```

Reply:
Also we could use `unknown` here.

Reply:
Updated to `delete<T>(token: HttpContextToken<T>)` to be consistent with others, TS fallback to the value being unknown anyway.

Reply:
@FDIM Any reason to keep this generic type parameter? For other methods it **is** surely necessary, but for `delete` is useless.